lambda/create-case/validation.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+var rePostcode = regexp.MustCompile(`^[A-Z0-9]{5,7}$`)
+var reDateOfBirth = regexp.MustCompile(`^[0-9]{4}-[0-9]{2}-[0-9]{2}$`)
+
+func validate(data Request) (bool, []Error) {
+	validationErrors := []Error{}
+
+	if data.Source == "" {
+		validationErrors = append(validationErrors, Error{
+			Source: "/source",
+			Detail: "required",
+		})
+	} else if data.Source != LpaSourceApplicant && data.Source != LpaSourcePhone {
+		validationErrors = append(validationErrors, Error{
+			Source: "/source",
+			Detail: fmt.Sprintf("must be %s or %s", LpaSourceApplicant, LpaSourcePhone),
+		})
+	}
+
+	if data.Type == "" {
+		validationErrors = append(validationErrors, Error{
+			Source: "/type",
+			Detail: "required",
+		})
+	} else if data.Type != LpaTypeHealthAndWelfare && data.Type != LpaTypePersonalAndFinance {
+		validationErrors = append(validationErrors, Error{
+			Source: "/type",
+			Detail: fmt.Sprintf("must be %s or %s", LpaTypeHealthAndWelfare, LpaTypePersonalAndFinance),
+		})
+	}
+
+	if data.Donor.Name == "" {
+		validationErrors = append(validationErrors, Error{
+			Source: "/donor/name",
+			Detail: "required",
+		})
+	}
+
+	if data.Donor.DateOfBirth == "" {
+		validationErrors = append(validationErrors, Error{
+			Source: "/donor/dob",
```

Comment:
For all of these validation errors, should the source just be like "donor's dob" so that the validation error messages will be more readable/friendly to end users?

Reply:
Since this isn't user facing I think that should be up to downstream services to organise the errors. For example, if they want to put the error next to the form element then the service needs a computer-readable string to identify where the error was. It also saves us baking-in language: they might use "applicant" or "recipient" rather than donor.

Reply:
Ah thats a fair point 👍🏼